Sources/LeafKit/LeafLexer.swift

```diff
@@ -46,19 +57,38 @@ struct TemplateSource {
     }
     
     mutating func readWhile(_ check: (Character) -> Bool) -> String? {
-        return readSliceWhile(check).flatMap { String($0) }
+        return readSliceWhile(pop: true, check).flatMap { String($0) }
+    }
+    
+    mutating func peekWhile(_ check: (Character) -> Bool) -> String? {
+        return peekSliceWhile(check).flatMap { String($0) }
     }
     
-    mutating func readSliceWhile(_ check: (Character) -> Bool) -> [Character]? {
+    mutating func popWhile(_ check: (Character) -> Bool) -> Int {
+        return readSliceWhile(pop: true, check)?.count ?? 0
+    }
+    
+    mutating func readSliceWhile(pop: Bool, _ check: (Character) -> Bool) -> [Character]? {
         var str = [Character]()
         while let next = peek() {
             guard check(next) else { return str }
-            pop()
+            if pop { self.pop() }
             str.append(next)
         }
         return str
     }
     
+    mutating func peekSliceWhile(_ check: (Character) -> Bool) -> [Character]? {
+        var str = [Character]()
+        var index = 0
+        while let next = peek(aheadBy: index) {
+            guard check(next) else { return str }
+            str.append(next)
+            index += 1
+        }
+        return str
+    }
+    
     func peek(aheadBy idx: Int = 0) -> Character? {
```

Comment:
Useful non-popping checking properties for being less aggressive about over-consuming the template stream